From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Tue, 4 Jan 2022 18:11:32 -0500
Subject: [PATCH] (VMP) ingredient matching cache

Code originally licensed under the MIT license and created for the VMP project: https://github.com/RelativityMC/VMP-fabric Copyright (c) 2021-2022 ishland

diff --git a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
index 8f35445477507bbec3c0cb1dccfd888316951595..c87d4857985c0dca19bd6fe1ecc076c09df02944 100644
--- a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
+++ b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
@@ -28,6 +28,9 @@ import net.minecraft.world.entity.player.StackedContents;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.ItemLike;
+import java.util.stream.Collectors; // JettPack
+import java.util.Set; // JettPack
+import java.util.HashSet; // JettPack
 
 public final class Ingredient implements Predicate<ItemStack> {
 
@@ -38,6 +41,8 @@ public final class Ingredient implements Predicate<ItemStack> {
     @Nullable
     private IntList stackingIds;
     public boolean exact; // CraftBukkit
+    private Set<Item> matchingItems = null; // JettPack
+    private boolean isEmptyMatch = false; // JettPack
 
     public Ingredient(Stream<? extends Ingredient.Value> entries) {
         this.values = (Ingredient.Value[]) entries.toArray((i) -> {
@@ -65,33 +70,23 @@ public final class Ingredient implements Predicate<ItemStack> {
         if (itemstack == null) {
             return false;
         } else {
-            this.dissolve();
-            if (this.itemStacks.length == 0) {
-                return itemstack.isEmpty();
-            } else {
-                ItemStack[] aitemstack = this.itemStacks;
-                int i = aitemstack.length;
-
-                for (int j = 0; j < i; ++j) {
-                    ItemStack itemstack1 = aitemstack[j];
-
-                    // CraftBukkit start
-                    if (this.exact) {
-                        if (itemstack1.getItem() == itemstack.getItem() && ItemStack.tagMatches(itemstack, itemstack1)) {
-                            return true;
-                        }
-
-                        continue;
-                    }
-                    // CraftBukkit end
-                    if (itemstack1.is(itemstack.getItem())) {
-                        return true;
-                    }
-                }
-
-                return false;
+            // JettPack start - VMP: ingredient matching cache
+            Set<Item> matchingItems = this.matchingItems;
+            boolean isEmptyMatch = this.isEmptyMatch;
+            if (matchingItems == null) {
+                matchingItems = this.matchingItems = Arrays.stream(this.values)
+                        .flatMap(entry -> entry.getItems().stream())
+                        .filter(itemstack1 -> !itemstack1.isEmpty())
+                        .map(ItemStack::getItem)
+                        .collect(Collectors.toCollection(HashSet::new));
+                isEmptyMatch = this.isEmptyMatch = this.matchingItems.isEmpty();
+            }
+            if (itemstack.isEmpty()) {
+                return isEmptyMatch;
             }
+            return matchingItems.contains(itemstack.getItem());
         }
+        // JettPack end
     }
 
     public IntList getStackingIds() {
